Feature/retry

mjkrause opened the pull request:
* has retry of (idempotent) GET request to check whether FC workspace exists
* added `test_retry` mock test in `test_standalone_utils.py`
* tested lambda with curl

mjkrause committed:
first pass

mjkrause committed:
first pass still (corrected typo)

mjkrause committed:
all tests pass, but not directly testing retry

mjkrause committed:
corrected comment and deleted unnecessary import

mjkrause committed:
added test to mock retry function with httpbin

mjkrause committed:
* wrapped retry in try / except block
* * return Response object with body to client
* created `test_standalone_utils.py` for special tests

mjkrause committed:
* corrected typo

mjkrause commented:
refers to original ticket #9

coverbeck commented on app.py:
```diff
@@ -40,6 +40,8 @@ def exportBag():
     manifest_io = ManifestIO(data, url, workspace, namespace, auth)
     # Check whether the workspace exists in this namespace, otherwise create it.
     r_check_workspace = manifest_io.workspace_exists()
+    if hasattr(r_check_workspace, 'body'):  # is Chalice Response object?
```
This is hard to understand. Are you saying if it doesn't time out, there is no body attribute? It seems like a fragile test... I think I would lean towards catching the exception.

coverbeck commented on chalicelib/utils.py:
```diff
@@ -44,13 +47,29 @@ def __init__(self, data, url, workspace, namespace, auth=None):
+        except Exception as x:
+            # after three retrys:
+            return Response(
+                body={'Request failed:': x.__class__.__name__},
```
What is this value typically?

It would be good to have a more meaningful message.

More importantly, I'm pretty sure this will return a 200 status code, which you definitely don't want. If there is an error you need to return an error to the client

mjkrause replied:
True, it returned a 200. I changed it such that it returns the RetryError message.

mjkrause committed:
* failed retry now returns a dictionary to the client with the RetryError message as one of the values

mjkrause committed:
* `workspace_exists` now returns response, containing status code
* if that status code is not 201 or 404 code execution will be aborted
in `manage_workspace`
* added in `manage_workspace` and `upload_files` in `utils.py`
 status code by overriding the default value (200)
* removed the `isinstance` check in `app.py` as it isn't needed (see
second bullet point)

mjkrause committed:
* deleted print command

mjkrause committed:
* corrected comment

mjkrause committed:
* put back try / except
* if retries are exhausted 500 as Response object status code

mjkrause commented:
I put the try / except back into `workspace_exists` and hard coded the status code (and returned it as a Response object), couldn't figure out a way to obtain a status code any other way. That's error code FC would return upon server error.

coverbeck commented:
I don't think you should return a hardcoded 500.

If you catch the RetryException you can glean the information; I hard-coded a case with a 404:

```
try:
  ...
catch RetryException as ex:
   print(ex)

HTTPConnectionPool(host='localhost', port=8000): Max retries exceeded with url: /foo (Caused by ResponseError('too many 404 error responses',))
```
However, that seems sort of fragile to have to parse the text. If there's no better way (requires some research), then maybe it's better to just do the retries yourself in a loop.

mjkrause committed:
* used HTTPError to return requests response object to client if
retries are exhausted
* added method `requests_response_to_chalice_Response` avoid repetition
when forming a Chalice Response object for returns in several methods
* wrote test to for that method

coverbeck commented on chalicelib/utils.py:
```diff
@@ -58,21 +59,22 @@ def workspace_exists(self):
-                            headers={'Content-Type': 'application/json'},
-                            status_code=500)
+            resp = s.get(url=url, headers=headers)
+            resp.raise_for_status()
```
You do not need the `try` any more, and should not call `resp.raise_for_status()`, you can just return the response directly.

Also, based on our earlier discussions, I thought you said you always wanted to return a Chalice response object from methods in utils.py; you are now returning a requests.Response object.

mjkrause replied:
sorry, I was mistaken. the truth is that I always return a requests response object to shove data back and forth from app.py to utils.py, except in cases of failure. this is a special case because we're passing it to another method, regardless of success or failure.

coverbeck commented on chalicelib/utils.py:
```diff
@@ -58,21 +59,22 @@ def workspace_exists(self):
+            resp = s.get(url=url, headers=headers)
+            resp.raise_for_status()
+            return resp
+        except HTTPError as err:
```
See previous comment; you don't need a `try` -- you are doing a throw, only to catch it, and then return it. You can save a step by just returning it.

mjkrause replied:
re. try, I removed it. strange though, I seemed to have been convinced that we need it. re. `raise_for_status`, it's actually slightly better to leave without (but it was required for try/except) as it will raise an exception for all 4xx/5xx. and thus would raise one if the workspace doesn't exist (404).

coverbeck commented on chalicelib/utils.py:
```diff
@@ -166,6 +155,20 @@ def _make_payload(self):
+    a Chalice Response object from it.
+    
+    :param response: (response object) from requests module 
+    :param status_code: (int) default is 200
```
There is no status_code parameter

mjkrause replied:
yes, good catch, I'll correct it before merging

mjkrause replied:
Hi Charles, I've made a few more small changes. Please take a look.

mjkrause committed:
* removed try / except from `workspace_exists` as it is not
needed
* removed `raised_for_status` in that method because we don't
want to raise an exception anymore without try/except
* corrected doc string in `requests_response_to_chalice_Response`

mjkrause committed:
* changed from assert to self.assertEqual in
`test_requests_response_to_chalice_Response`
* removed unnecessary argument

coverbeck commented on chalicelib/utils.py:
```diff
@@ -44,17 +45,32 @@ def __init__(self, data, url, workspace, namespace, auth=None):
-        """Returns true if FireCloud workspace in the specified namespace
-        exists, otherwise false.
-        """
+        """Returns requests response object. Status code is 200 if FireCloud 
```
There is a whole range of response codes that can come back, not just 200 and 404.

coverbeck reviewed (approved):
One comment about your comment; not crucial
